File: sandbox/src/sandbox.rs
use std::collections::HashMap;
use std::io::Read;
use std::os::unix::process::CommandExt;
use std::path::PathBuf;
use std::process::Command as StdCommand;
use std::process::Stdio;
use std::time::{Duration, Instant};

use tonic::{Request, Response, Status};
use tracing::info;

// ── Generated protobuf types ────────────────────────────────
tonic::include_proto!("sandbox");

pub use sandbox_service_server::SandboxServiceServer;

// ── Resource limits via setrlimit ────────────────────────────
fn apply_resource_limits(memory_bytes: u64, _cpu_shares: u64, disk_bytes: u64, max_pids: u32) {
    unsafe {
        if memory_bytes > 0 {
            let rlim = libc::rlimit {
                rlim_cur: memory_bytes as libc::rlim_t,
                rlim_max: memory_bytes as libc::rlim_t,
            };
            libc::setrlimit(libc::RLIMIT_AS, &rlim);
        }
        if disk_bytes > 0 {
            let rlim = libc::rlimit {
                rlim_cur: disk_bytes as libc::rlim_t,
                rlim_max: disk_bytes as libc::rlim_t,
            };
            libc::setrlimit(libc::RLIMIT_FSIZE, &rlim);
        }
        if max_pids > 0 {
            let rlim = libc::rlimit {
                rlim_cur: max_pids as libc::rlim_t,
                rlim_max: max_pids as libc::rlim_t,
            };
            libc::setrlimit(libc::RLIMIT_NPROC, &rlim);
        }
    }
}

// ── Sandbox service implementation ──────────────────────────
#[derive(Default)]
pub struct SandboxServiceImpl;

impl SandboxServiceImpl {
    fn sandbox_dir(task_id: &str, node_id: &str) -> std::io::Result<PathBuf> {
        let dir = std::env::temp_dir()
            .join("lingxi-sandbox")
            .join(format!("{}-{}", task_id, node_id));
        std::fs::create_dir_all(&dir)?;
        Ok(dir)
    }

    fn write_input_files(dir: &PathBuf, files: &HashMap<String, Vec<u8>>) -> std::io::Result<()> {
        for (name, content) in files {
            let path = dir.join(name);
            if let Some(parent) = path.parent() {
                std::fs::create_dir_all(parent)?;
            }
            std::fs::write(&path, content)?;
        }
        Ok(())
    }

    fn cleanup(dir: &PathBuf) {
        let _ = std::fs::remove_dir_all(dir);
    }
}

#[tonic::async_trait]
impl sandbox_service_server::SandboxService for SandboxServiceImpl {
    async fn execute(
        &self,
        request: Request<ExecuteRequest>,
    ) -> Result<Response<ExecuteResult>, Status> {
        let req = request.into_inner();
        let task_id = req.task_id.clone();
        let node_id = req.node_id.clone();

        info!(
            task_id = %task_id,
            node_id = %node_id,
            command = %req.command,
            "executing in sandbox"
        );

        // Create isolated sandbox directory
        let work_dir = Self::sandbox_dir(&task_id, &node_id)
            .map_err(|e| Status::internal(format!("failed to create sandbox dir: {}", e)))?;

        // Write input files
        if !req.input_files.is_empty() {
            Self::write_input_files(&work_dir, &req.input_files)
                .map_err(|e| Status::internal(format!("failed to write input files: {}", e)))?;
        }

        // Parse limits
        let mem_limit = req.limits.as_ref().map(|l| l.memory_bytes).unwrap_or(0);
        let cpu_shares = req.limits.as_ref().map(|l| l.cpu_shares).unwrap_or(0);
        let disk_limit = req.limits.as_ref().map(|l| l.disk_bytes).unwrap_or(0);
        let pid_limit = req.limits.as_ref().map(|l| l.max_pids).unwrap_or(0);

        let timeout_sec = if req.timeout_sec > 0 { req.timeout_sec } else { 30 };
        let env_map = req.env;
        let work_dir_clone = work_dir.clone();

        // Execute in blocking thread (setrlimit must run in child process)
        let start = Instant::now();
        let result = tokio::task::spawn_blocking(move || -> (i32, Vec<u8>, Vec<u8>, bool, String) {
            let mut cmd = StdCommand::new(&req.command);
            cmd.args(&req.args)
                .current_dir(&work_dir_clone)
                .envs(env_map.into_iter())
                .stdin(Stdio::null())
                .stdout(Stdio::piped())
                .stderr(Stdio::piped());

            unsafe {
                cmd.pre_exec(move || {
                    apply_resource_limits(mem_limit, cpu_shares, disk_limit, pid_limit);
                    Ok(())
                });
            }

            let mut child = match cmd.spawn() {
                Ok(c) => c,
                Err(e) => return (-1, vec![], vec![], false, format!("spawn failed: {}", e)),
            };

            // Drain stdout/stderr in background threads to prevent pipe deadlock
            let stdout_handle = child.stdout.take().map(|mut out| {
                std::thread::spawn(move || {
                    let mut buf = Vec::new();
                    let _ = out.read_to_end(&mut buf);
                    buf
                })
            });

            let stderr_handle = child.stderr.take().map(|mut err| {
                std::thread::spawn(move || {
                    let mut buf = Vec::new();
                    let _ = err.read_to_end(&mut buf);
                    buf
                })
            });

            // Poll for completion with timeout
            let now = Instant::now();
            let status = loop {
                if now.elapsed().as_secs() >= timeout_sec as u64 {
                    let _ = child.kill();
                    break (None, true);
                }
                match child.try_wait() {
                    Ok(Some(status)) => break (Some(status), false),
                    Ok(None) => std::thread::sleep(Duration::from_millis(50)),
                    Err(e) => return (-1, vec![], vec![], false, format!("wait failed: {}", e)),
                }
            };

            let stdout = stdout_handle
                .and_then(|h| h.join().ok())
                .unwrap_or_default();
            let stderr = stderr_handle
                .and_then(|h| h.join().ok())
                .unwrap_or_default();

            match status {
                (Some(s), timed_out) => (s.code().unwrap_or(-1), stdout, stderr, timed_out, String::new()),
                (None, timed_out) => (-1, stdout, stderr, timed_out, "process killed".into()),
            }
        }).await;

        let elapsed = start.elapsed();

        let (exit_code, stdout, stderr, timed_out, error_msg) = match result {
            Ok(r) => r,
            Err(e) => (-1, vec![], vec![], false, format!("task join error: {}", e)),
        };

        // Cleanup sandbox directory
        Self::cleanup(&work_dir);

        let resource_usage = ResourceUsage {
            memory_peak_bytes: 0,
            cpu_time_usec: elapsed.as_micros() as u64,
            user_time_usec: 0,
            system_time_usec: 0,
        };

        info!(
            task_id = %task_id,
            node_id = %node_id,
            exit_code = exit_code,
            duration_ms = elapsed.as_millis() as u64,
            timed_out = timed_out,
            "sandbox execution complete"
        );

        Ok(Response::new(ExecuteResult {
            exit_code: exit_code as i32,
            stdout,
            stderr,
            timed_out,
            resource_usage: Some(resource_usage),
            error: error_msg,
        }))
    }
}

File: aios-core/internal/worker/executor/sandbox.go
package executor

import (
	"context"
	"fmt"
	"time"

	"google.golang.org/grpc"
	"google.golang.org/grpc/credentials/insecure"

	"github.com/tangying-ai/aios-core/internal/worker/executor/sandboxpb"
)

// SandboxExecutor executes commands via the Rust sandbox gRPC service.
// Provides resource isolation (memory, CPU, disk, PID limits), timeout enforcement,
// and automatic cleanup of temporary execution directories.
type SandboxExecutor struct {
	address string
	conn    *grpc.ClientConn
	client  sandboxpb.SandboxServiceClient
}

func NewSandboxExecutor(address string) (*SandboxExecutor, error) {
	conn, err := grpc.NewClient(address,
		grpc.WithTransportCredentials(insecure.NewCredentials()),
		grpc.WithDefaultCallOptions(grpc.MaxCallRecvMsgSize(32*1024*1024)),
	)
	if err != nil {
		return nil, fmt.Errorf("failed to create sandbox gRPC client: %w", err)
	}

	return &SandboxExecutor{
		address: address,
		conn:    conn,
		client:  sandboxpb.NewSandboxServiceClient(conn),
	}, nil
}

func (s *SandboxExecutor) Close() error {
	if s.conn != nil {
		return s.conn.Close()
	}
	return nil
}

func (s *SandboxExecutor) Execute(ctx context.Context, req ExecutionRequest) (ExecutionResult, error) {
	if s.client == nil {
		return ExecutionResult{Error: "sandbox client not initialized"}, nil
	}

	// Map ResourceLimits to protobuf
	var pbLimits *sandboxpb.ResourceLimits
	if req.Limits.MemoryBytes > 0 || req.Limits.CPUShares > 0 || req.Limits.DiskBytes > 0 || req.Limits.MaxPIDs > 0 {
		pbLimits = &sandboxpb.ResourceLimits{
			MemoryBytes: req.Limits.MemoryBytes,
			CpuShares:   req.Limits.CPUShares,
			DiskBytes:   req.Limits.DiskBytes,
			MaxPids:     req.Limits.MaxPIDs,
		}
	}

	// Convert env map
	envMap := make(map[string]string)
	for k, v := range req.Env {
		envMap[k] = v
	}

	// Build request
	pbReq := &sandboxpb.ExecuteRequest{
		TaskId:     req.TaskID,
		NodeId:     req.NodeID,
		Command:    req.Command,
		Args:       req.Args,
		Env:        envMap,
		WorkDir:    req.WorkDir,
		TimeoutSec: req.TimeoutSec,
		Limits:     pbLimits,
		InputFiles: req.InputFiles,
		Stdin:      req.Stdin,
	}

	// Execute with context timeout
	execCtx, cancel := context.WithTimeout(ctx, time.Duration(req.TimeoutSec+5)*time.Second)
	defer cancel()

	pbResult, err := s.client.Execute(execCtx, pbReq)
	if err != nil {
		return ExecutionResult{Error: fmt.Sprintf("sandbox gRPC call failed: %v", err)}, nil
	}

	result := ExecutionResult{
		ExitCode: pbResult.ExitCode,
		Stdout:   pbResult.Stdout,
		Stderr:   pbResult.Stderr,
		TimedOut: pbResult.TimedOut,
		Error:    pbResult.Error,
	}

	if pbResult.ResourceUsage != nil {
		result.ResourceUsage = &ResourceUsage{
			MemoryPeakBytes: pbResult.ResourceUsage.MemoryPeakBytes,
			CPUTimeUsec:     pbResult.ResourceUsage.CpuTimeUsec,
			UserTimeUsec:    pbResult.ResourceUsage.UserTimeUsec,
			SystemTimeUsec:  pbResult.ResourceUsage.SystemTimeUsec,
		}
	}

	return result, nil
}
